workspaces/optic-engine/src/projections/spectacle/endpoints.rs

```diff
@@ -344,28 +349,123 @@ impl EndpointsProjection {
     }
   }
 
+  ////////////////////////////////////////////////////////////////////////////////////////////////////
+  fn get_or_create_endpoint_node_index(
```

Comment:
Nit: we've called these get or create methods "ensure_" in the other projections, could be nice to have that consistent :)